src/aegis/src/aegis_server.erl

```diff
@@ -0,0 +1,271 @@
+
+
+-spec encrypt(Db :: #{}, Key :: binary(), Value :: binary()) -> binary().
+encrypt(#{} = Db, Key, Value) when is_binary(Key), is_binary(Value) ->
+    #{
+        uuid := UUID
+    } = Db,
+
+    case ets:member(?KEY_CHECK, UUID) of
+        true ->
+            gen_server:call(?MODULE, {encrypt, Db, Key, Value});
+        false ->
+            process_flag(sensitive, true),
+
+            {ok, DbKey} = do_open_db(Db),
+            do_encrypt(DbKey, Db, Key, Value)
+    end.
+
+
+-spec decrypt(Db :: #{}, Key :: binary(), Value :: binary()) -> binary().
+decrypt(#{} = Db, Key, Value) when is_binary(Key), is_binary(Value) ->
+    #{
+        uuid := UUID
+    } = Db,
+
+    case ets:member(?KEY_CHECK, UUID) of
+        true ->
+            gen_server:call(?MODULE, {decrypt, Db, Key, Value});
+        false ->
+            process_flag(sensitive, true),
+
+            {ok, DbKey} = do_open_db(Db),
+            do_decrypt(DbKey, Db, Key, Value)
+    end.
+
+
+%% gen_server functions
+
+init([]) ->
+    process_flag(sensitive, true),
+    Cache = ets:new(?MODULE, [set, private, {keypos, #entry.uuid}]),
+    ets:new(?KEY_CHECK, [named_table, protected, {read_concurrency, true}]),
+
+    St = #{
+        cache => Cache
+    },
+    {ok, St, ?INIT_TIMEOUT}.
+
+
+terminate(_Reason, _St) ->
+    ok.
+
+
+handle_call({insert_key, UUID, DbKey}, _From, #{cache := Cache} = St) ->
+    ok = insert(Cache, UUID, DbKey),
+    {reply, ok, St, ?TIMEOUT};
+
+handle_call({encrypt, #{uuid := UUID} = Db, Key, Value}, From, St) ->
+    #{
+        cache := Cache
+    } = St,
+
+    {ok, DbKey} = lookup(Cache, UUID),
+
+    erlang:spawn(fun() ->
+        process_flag(sensitive, true),
+        try
+            do_encrypt(DbKey, Db, Key, Value)
+        of
+            Resp ->
+                gen_server:reply(From, Resp)
+        catch
+            _:Error ->
+                gen_server:reply(From, {error, Error})
+        end
+    end),
+
+    {noreply, St, ?TIMEOUT};
+
+handle_call({decrypt, #{uuid := UUID} = Db, Key, Value}, From, St) ->
+    #{
+        cache := Cache
+    } = St,
+
+    {ok, DbKey} = lookup(Cache, UUID),
+
+    erlang:spawn(fun() ->
+        process_flag(sensitive, true),
+        try
+            do_decrypt(DbKey, Db, Key, Value)
+        of
+            Resp ->
+                gen_server:reply(From, Resp)
+        catch
+            _:Error ->
+                gen_server:reply(From, {error, Error})
+        end
+    end),
+
+    {noreply, St, ?TIMEOUT};
+
+handle_call(_Msg, _From, St) ->
+    {noreply, St}.
+
+
+handle_cast(_Msg, St) ->
+    {noreply, St}.
+
+
+handle_info(_Msg, St) ->
+    {noreply, St}.
+
+
+code_change(_OldVsn, St, _Extra) ->
+    {ok, St}.
+
+
+%% private functions
+
+do_open_db(#{} = Db) ->
+    #{
+        uuid := UUID,
+        user_ctx := UserCtx,
+        db_options := Options0
+    } = Db,
+
+    %% put back elements removed in fabric2_fdb:open/2
+    Options = [{uuid, UUID}, {user_ctx, UserCtx} | Options0],
```

Comment:
This doesn't appear to be necessary. both items are present by the time aegis:open_db is called. If it _is_ necessary, we should change the far end (fabric2_fdb:open) rather than depend on the current behaviour of its implementation.

Reply:
This is subtle, this method not called by `aegis_server:open_db/2`, it called by `encrypt/decrypt/2` when they are re-retriving db key - they don't have `Options` available and have to restore it.

I can rename this method to something like `do_reopen_db/1` to clarify the intent and use-case.